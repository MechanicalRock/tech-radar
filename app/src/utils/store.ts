import * as reducers from '../reducers';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import thunk from 'redux-thunk';

const store = createStore(
  combineReducers<ReduxState>(reducers),
  applyMiddleware(thunk)
);

export default store;
