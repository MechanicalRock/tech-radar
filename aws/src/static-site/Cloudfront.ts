import * as s3 from '@aws-cdk/aws-s3';
import { Stack } from '@aws-cdk/cdk';
import {
  CloudFrontWebDistribution,
  CloudFrontWebDistributionProps,
  CfnCloudFrontOriginAccessIdentity,
  PriceClass,
  CloudFrontAllowedMethods,
  CloudFrontAllowedCachedMethods
} from '@aws-cdk/aws-cloudfront';
import { Bucket } from '@aws-cdk/aws-s3';
import { Output } from '@aws-cdk/cdk';

export interface CloudfrontProps {
  stack: Stack;
  logicalName: string;
  siteBucket: s3.Bucket;
  originAccessIdentity: CfnCloudFrontOriginAccessIdentity;
}

export class Cloudfront {
  distribution: CloudFrontWebDistribution;

  constructor(private props: CloudfrontProps) {
    this.distribution = this.create();
    this.createOutputs();
  }

  private create(): CloudFrontWebDistribution {
    const distributionProps: CloudFrontWebDistributionProps = {
      comment: 'Cloud front distribution for static website',
      aliasConfiguration: {
        acmCertRef:
          'arn:aws:acm:us-east-1:598112752826:certificate/da98edf4-7dc7-4ee8-95d6-82d96bc1a13b',
        names: ['techradar.cloud', 'www.techradar.cloud']
      },
      originConfigs: [
        {
          s3OriginSource: {
            s3BucketSource: this.props.siteBucket,
            originAccessIdentity: this.props.originAccessIdentity
          },
          behaviors: [
            {
              isDefaultBehavior: true,
              compress: true,
              allowedMethods: CloudFrontAllowedMethods.GET_HEAD_OPTIONS,
              cachedMethods: CloudFrontAllowedCachedMethods.GET_HEAD_OPTIONS,
              minTtlSeconds: 3600
            }
          ]
        }
      ],
      errorConfigurations: [
        {
          errorCode: 404,
          responseCode: 200,
          responsePagePath: '/'
        }
      ],
      loggingConfig: {
        bucket: this.createLoggingBucket()
      },
      priceClass: PriceClass.PriceClassAll
    };
    return new CloudFrontWebDistribution(
      this.props.stack,
      this.props.logicalName,
      distributionProps
    );
  }

  private createOutputs() {
    new Output(this.props.stack, 'DistributionDomainName', {
      value: this.distribution.domainName,
      disableExport: true
    });
  }

  private createLoggingBucket(): Bucket {
    return new Bucket(this.props.stack, 'LogsBucket', {
      encryption: s3.BucketEncryption.S3Managed
    });
  }
}
