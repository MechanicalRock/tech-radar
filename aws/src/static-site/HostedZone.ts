import { Stack } from '@aws-cdk/cdk';
import {
  AliasRecord,
  IAliasRecordTarget,
  HostedZone
} from '@aws-cdk/aws-route53';

export interface HostedZoneProps {
  stack: Stack;
  fullyQualifiedDomainName: string;
}

export class HostedZoneAlias {
  hostedZone: HostedZone;
  alias: IAliasRecordTarget;

  constructor(private props: HostedZoneProps, alias: IAliasRecordTarget) {
    this.alias = alias;
    this.create();
  }

  create(): void {
    this.hostedZone = new HostedZone(this.props.stack, 'HostedZone', {
      zoneName: this.props.fullyQualifiedDomainName,
      comment: `Hosted zone for domain: ${this.props.fullyQualifiedDomainName}`
    });

    new AliasRecord(this.props.stack, 'CloudFrontAlias', {
      zone: this.hostedZone,
      recordName: this.props.fullyQualifiedDomainName,
      target: this.alias
    });
  }
}
