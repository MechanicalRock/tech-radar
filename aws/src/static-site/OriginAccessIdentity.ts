import { Stack } from '@aws-cdk/cdk';
import { CfnCloudFrontOriginAccessIdentity } from '@aws-cdk/aws-cloudfront';

export class OriginAccessIdentity {
  private originAccessIdentity: CfnCloudFrontOriginAccessIdentity;
  constructor(private stack: Stack) {
    this.originAccessIdentity = this.create();
  }

  getResource(): CfnCloudFrontOriginAccessIdentity {
    return this.originAccessIdentity;
  }

  private create(): CfnCloudFrontOriginAccessIdentity {
    return new CfnCloudFrontOriginAccessIdentity(
      this.stack,
      'OriginAccessIdentity',
      {
        cloudFrontOriginAccessIdentityConfig: {
          comment:
            'Cloud front user which will be allowed to access the site s3 bucket'
        }
      }
    );
  }
}
