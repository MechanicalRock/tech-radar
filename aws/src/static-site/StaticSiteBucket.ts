import { Stack, Parameter } from '@aws-cdk/cdk';
import * as s3 from '@aws-cdk/aws-s3';
import { PolicyStatement, CanonicalUserPrincipal } from '@aws-cdk/aws-iam';

export interface StaticSiteBucketProps {
  stack: Stack;
  logicalName: string;
  originAccessIdentityCanonicalId: string;
}

export class StaticSiteBucket {
  bucketParam: Parameter;
  bucket: s3.Bucket;

  constructor(private props: StaticSiteBucketProps) {
    this.bucket = this.create();
  }

  private create() {
    const bucket = new s3.Bucket(this.props.stack, this.props.logicalName, {
      encryption: s3.BucketEncryption.S3Managed
    });

    const statement = new PolicyStatement();
    statement.addPrincipal(
      new CanonicalUserPrincipal(this.props.originAccessIdentityCanonicalId)
    );
    statement.addAction('s3:GetObject');
    statement.addResource(bucket.bucketArn + '/*');
    bucket.addToResourcePolicy(statement);
    return bucket;
  }
}
