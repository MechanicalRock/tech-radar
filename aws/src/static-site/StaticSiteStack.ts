import * as cdk from '@aws-cdk/cdk';
import { Parameter } from '@aws-cdk/cdk';
import { StaticSiteBucket } from './StaticSiteBucket';
import { OriginAccessIdentity } from './OriginAccessIdentity';
import { Cloudfront } from './Cloudfront';
import { HostedZoneAlias } from './HostedZone';

export class StaticSiteStack extends cdk.Stack {
  bucketParameter: Parameter;

  constructor(parent: cdk.App, stackName: string, props?: cdk.StackProps) {
    super(parent, stackName, props);
    this.create();
  }

  create(): void {
    const originAccessIdentityResource = new OriginAccessIdentity(
      this
    ).getResource();

    const siteBucket = new StaticSiteBucket({
      stack: this,
      logicalName: 'StaticSiteBucket',
      originAccessIdentityCanonicalId:
        originAccessIdentityResource.cloudFrontOriginAccessIdentityS3CanonicalUserId
    });

    const { distribution } = new Cloudfront({
      stack: this,
      siteBucket: siteBucket.bucket,
      logicalName: 'static-site-cloudfront',
      originAccessIdentity: originAccessIdentityResource
    });

    new HostedZoneAlias(
      {
        stack: this,
        fullyQualifiedDomainName: 'techradar.cloud'
      },
      distribution
    );
  }
}
