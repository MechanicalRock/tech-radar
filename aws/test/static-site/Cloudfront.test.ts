import { Stack } from '@aws-cdk/cdk';
import * as yaml from '../utils/Yaml';
import { Cloudfront } from '../../src/static-site/Cloudfront';
import { Bucket } from '@aws-cdk/aws-s3';
import { OriginAccessIdentity } from '../../src/static-site/OriginAccessIdentity';

describe('Cloudfront', () => {
    it('should match the template snapshot', () => {
        // Given
        
        const cloudfrontStack = new Stack();
        const siteBucket = new Bucket(cloudfrontStack, 'MySiteBucket');
        const originAccessIdentity = new OriginAccessIdentity(cloudfrontStack);
        
        // When
        new Cloudfront({
            stack: cloudfrontStack,
            logicalName: 'MyCloudfront',
            siteBucket,
            originAccessIdentity: originAccessIdentity.getResource()
        });

        // Then
        expect(yaml.toYamlString(cloudfrontStack.toCloudFormation())).toMatchSnapshot();
    })
})