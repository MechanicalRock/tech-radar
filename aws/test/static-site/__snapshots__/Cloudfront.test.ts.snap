// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloudfront should match the template snapshot 1`] = `
"Resources:
  MySiteBucket1481A238:
    Type: 'AWS::S3::Bucket'
    DeletionPolicy: Retain
  OriginAccessIdentity:
    Type: 'AWS::CloudFront::CloudFrontOriginAccessIdentity'
    Properties:
      CloudFrontOriginAccessIdentityConfig:
        Comment: Cloud front user which will be allowed to access the site s3 bucket
  LogsBucket9C4D8843:
    Type: 'AWS::S3::Bucket'
    Properties:
      BucketEncryption:
        ServerSideEncryptionConfiguration:
          - ServerSideEncryptionByDefault:
              SSEAlgorithm: AES256
    DeletionPolicy: Retain
  MyCloudfrontCFDistribution4F715D75:
    Type: 'AWS::CloudFront::Distribution'
    Properties:
      DistributionConfig:
        Aliases:
          - techradar.cloud
          - www.techradar.cloud
        CacheBehaviors: []
        Comment: Cloud front distribution for static website
        CustomErrorResponses:
          - ErrorCode: 404
            ResponseCode: 200
            ResponsePagePath: /
        DefaultCacheBehavior:
          AllowedMethods:
            - GET
            - HEAD
            - OPTIONS
          CachedMethods:
            - GET
            - HEAD
            - OPTIONS
          Compress: true
          ForwardedValues:
            Cookies:
              Forward: none
            QueryString: false
          MinTTL: 3600
          TargetOriginId: origin1
          ViewerProtocolPolicy: redirect-to-https
        DefaultRootObject: index.html
        Enabled: true
        HttpVersion: http2
        IPV6Enabled: true
        Logging:
          Bucket:
            'Fn::GetAtt':
              - LogsBucket9C4D8843
              - DomainName
          IncludeCookies: false
        Origins:
          - DomainName:
              'Fn::GetAtt':
                - MySiteBucket1481A238
                - DomainName
            Id: origin1
            S3OriginConfig:
              OriginAccessIdentity:
                'Fn::Join':
                  - ''
                  - - origin-access-identity/cloudfront/
                    - Ref: OriginAccessIdentity
        PriceClass: PriceClass_All
        ViewerCertificate:
          AcmCertificateArn: >-
            arn:aws:acm:us-east-1:598112752826:certificate/da98edf4-7dc7-4ee8-95d6-82d96bc1a13b
          SslSupportMethod: sni-only
Outputs:
  DistributionDomainName:
    Value:
      'Fn::GetAtt':
        - MyCloudfrontCFDistribution4F715D75
        - DomainName
"
`;
